lib/pure/os.nim

```diff
@@ -1687,6 +1687,34 @@ proc setFilePermissions*(filename: string, permissions: set[FilePermission]) {.
       var res2 = setFileAttributesA(filename, res)
     if res2 == - 1'i32: raiseOSError(osLastError(), $(filename, permissions))
 
+template fromUnixImpl(unixPerm: range[0..7]; i: static[int]): set[FilePermission] =
+  var r = fpUserRead
+  var w = fpUserWrite
+  var x = fpUserExec
+  when i > 0:
+    inc r, i
+    inc w, i
+    inc x, i
+  case unixPerm
+  of 0: {}
+  of 1: {x}
+  of 2: {w}
+  of 3: {w, x}
+  of 4: {r}
+  of 5: {r, x}
+  of 6: {r, w}
+  of 7: {r, w, x}
+
+func fromUnixFilePermission*(user: range[0..7]; group: range[0..7]; other: range[0..7]): set[FilePermission] {.inline, since: (1, 3).} =
+  ## Convenience func to convert Unix like file permission to ``set[FilePermission]``.
+  ##
+  ## See also:
+  ## * `getFilePermissions <#getFilePermissions,string>`_
+  ## * `setFilePermissions <#setFilePermissions,string,set[FilePermission]>`_
+  runnableExamples:
+    static: doAssert fromUnixFilePermission(7, 7, 7) == {fpUserExec, fpUserWrite, fpUserRead, fpGroupExec, fpGroupWrite, fpGroupRead, fpOthersExec, fpOthersWrite, fpOthersRead}
+  result = fromUnixImpl(user, 0) + fromUnixImpl(group, 3) + fromUnixImpl(other, 6)
```

Comment:
You should instead take in the permission number in octal form: `0o700` or similar, then just do some simple bitwise ops to get the actual value. That'd be more useful as you can then use this to decode user input.

You can see here for a quick description of how octal permission values are represented: https://unix.stackexchange.com/a/9514

Reply:
Hrm, I think I'm against this - octal numbers aren't used very often, and people are used to `chmod DDD` (no octal prefix there).

As an example, Ansible [does this](https://docs.ansible.com/ansible/latest/modules/file_module.html) and they have to explicitly call out in their documentation that the octal prefix has to be used.

Reply:
In the end octal numbers are just integers, an application would simply use `strutils.parseOctInt` on the string and send it to this proc for decoding.

The idea here is to take a whole integer instead of separated numbers, which would be more efficient and can be easily stored as a state variable to pass around. This also have the benefits of allowing users to decode `stat()` output and get Nim's set representation of it.

Reply:
yes, can we have `let perm = fromUnixFilePermission 0o644` ?

Reply:
> a way to also parse update permissions such as: `"ug-x"` (as in `chmod ug-x foo`)

That arguably should not be in the stdlib :P